homeassistant/components/home_connect/diagnostics.py

```diff
@@ -23,3 +25,14 @@ async def async_get_config_entry_diagnostics(
         }
         for device in hass.data[DOMAIN][config_entry.entry_id].devices
     }
+
+
+async def async_get_device_diagnostics(
+    hass: HomeAssistant, config_entry: ConfigEntry, device: DeviceEntry
+) -> dict[str, Any]:
+    """Return diagnostics for a device."""
+    appliance = _get_appliance_by_device_id(hass, device.id)
+    return {
+        "status": appliance.status,
+        "programs": await hass.async_add_executor_job(appliance.get_programs_available),
+    }
```

Comment:
So I have two concerns now.
My first concern is that I think it would make sense to use a module function to reduce the duplicate code with `async_get_config_entry_diagnostics`
My second concern is that I don't think it's a good idea to have `hass.async_add_executor_job` inside a loop.

So I think it would make sense to create two function like this:

```python
def _generate_appliance_diagnostics(appliance):
    return {
        "status": appliance.status,
        "programs": appliance.get_programs_available(),
    }

def _generate_entry_diagnostics(devices):
    return {
        device.appliance.haId: _generate_appliance_diagnostics(device.appliance)
        for device in hass.data[DOMAIN][config_entry.entry_id].devices
    }
```

Reply:
You would keep `hass.async_add_executor_job` in the main function (not in the child function)

```python
async def async_get_config_entry_diagnostics(
    hass: HomeAssistant, config_entry: ConfigEntry
) -> dict[str, Any]:
    """Return diagnostics for a config entry."""
    return await hass.async_add_executor_job(
        _generate_entry_diagnostics(
            hass.data[DOMAIN][config_entry.entry_id].devices
        )
    )
```

Reply:
Okey, that makes a lot of sense